### PR TITLE
modem: verify send semaphore before taking any action and fix race condition waiting on reply

### PR DESCRIPTION
**Pull request by mniestroj**
Check function arguments on the beginning of function, so that -EINVAL is
returned before taking any actions, such as sending data over modem interface.

Fix race condition, by first reseting semaphore, then sending command over modem
interface, then waiting on reply by taking semaphore.

Add minor improvement to handling error when setting up command handlers - there
is no need to clear them out in that case.

### CODE REVIEW

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by pfalcon** [comment]
@mniestroj: Changes like these are nice, though reviewing them apparently better be handled by people familiar with "modem" infrastructure. I just noticed that `tests/` doesn't have any directory or C file matching "*modem*" pattern, which likely translates that there're no tests for the modem infra. Did you consider doing something about that?

**Comment by pfalcon** [comment]
And the esp change apparently better belongs to its own PR.

**Comment by mniestroj** [comment]
>I just noticed that tests/ doesn't have any directory or C file matching "modem" pattern, which likely translates that there're no tests for the modem infra. Did you consider doing something about that?

Unfortunately. I didn't consider writing unit tests for modem infra, but it seems a very good idea. The issue is that I don't have time for such time demanding task now, because it involves writing tests from scratch. For now I am just trying to fix stuff here and there to make things work reliably with ESP WiFi driver. I will keep this in mind when working on another modem related driver.

>And the esp change apparently better belongs to its own PR.

I don't agree here. This change is just removing workaround for the issue that is addressed in this PR. esp change is trivial and doesn't bloat this PR, so personally I think it suits in this PR quite well.

**Review by nashif — APPROVED** [review]

File: drivers/modem/modem_cmd_handler.c [hunk 1]
```diff
@@ -476,7 +476,15 @@ static int _modem_cmd_send(struct modem_iface *iface,
...
+		return -EINVAL;
+	}
+
+	if (K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
+		/* semaphore is not needed if there is no timeout */
+		sem = NULL;
+	} else if (!sem) {
+		/* cannot respect timeout without semaphore */
 		return -EINVAL;
 	}
```
Comment (WilliamGFish) [inline drivers/modem/modem_cmd_handler.c:489]: The first evaluation of the semaphore (ln 479) always fails if timeout is used when no sem is sent.

Reworking the line to:
` if (!iface || !handler || !handler->cmd_handler_data || !buf )`


Billy...

Comment (WilliamGFish) [reply drivers/modem/modem_cmd_handler.c:489]: The UBlox sara-modem now fails

**Review by WilliamGFish — COMMENTED** [review]
This causes modems to stop working

File: drivers/modem/modem_cmd_handler.c (hunk 1 above)
Comment (mniestroj) [reply drivers/modem/modem_cmd_handler.c:489]: @WilliamGFish You are right, `!sem` shouldn't be there. I left it there by mistake... Will you send a PR?

Comment (XavierChapron) [reply drivers/modem/modem_cmd_handler.c:489]: @WilliamGFish I do agree. This is introducing an error when called without sem nor timeout [here](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/modem/ublox-sara-r4.c#L365).

Comment (mniestroj) [reply drivers/modem/modem_cmd_handler.c:489]: #30458

**Pull request merged**